### PR TITLE
[Docs] Add docstring for `MMArchitectureQuant` & `NativeQuantizer`

### PR DESCRIPTION
**Pull request by 415905716**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmrazor/models/algorithms/quantization/mm_architecture.py [hunk 1]
```diff
@@ -97,11 +117,18 @@ def traverse(module, prefix):
 
         src_state_dict = self.qmodels[src_mode].state_dict()
         for mode in self.forward_modes:
+            import pdb
+            pdb.set_trace()
```
Comment (pppppM) [inline mmrazor/models/algorithms/quantization/mm_architecture.py:121]: del

File: mmrazor/models/algorithms/quantization/mm_architecture.py [hunk 2]
```diff
@@ -28,31 +28,39 @@ class MMArchitectureQuant(BaseAlgorithm):
...
         qmodel_modes (list): The available mode of runner.
         data_preprocessor (dict | torch.nn.Module | None): The pre-process
             config of :class:`BaseDataPreprocessor`. Defaults to None.
-        pretrained_ckpt (str, Optional): The path of pretrained checkpoint.
-            Defaults to None.
-        init_cfg (dict): The weight initialized config for
-            :class:`BaseModule`.
+        forward_modes (tuple): The modes in forward method in OpenMMLab
+            architecture could be tensor, predict, or loss. It can generate
+            different graph of quantized model.
```
Comment (pppppM) [inline mmrazor/models/algorithms/quantization/mm_architecture.py:36]: Add a note to explain why we need to generate multi graph

File: mmrazor/models/quantizers/native_quantizer.py [hunk 3]
```diff
@@ -80,6 +108,16 @@ def support_a_modes(self):
...
+
+        Args:
+            model (_type_): _description_
+            graph_module (_type_): Graph modules before fuse.
+
+        Returns:
+            ObserverdGraphModule: Graph module after fuse and observerd
+        """
+
         graph_module = _fuse_fx(
```
Comment (pppppM) [inline mmrazor/models/quantizers/native_quantizer.py:121]: Explain what happened after `_fuse_fx` and after `prepare`

File: mmrazor/models/quantizers/native_quantizer.py [hunk 4]
```diff
@@ -96,16 +134,30 @@ def prepare(self, model, graph_module):
 
     def post_process_weight_fakequant(self,
                                       observed_module,
-                                      keep_fake_quant=False):
+                                      keep_fake_quant: bool = False):
+        """weight fakequant for supported QAT modules.
+
```
Comment (pppppM) [inline mmrazor/models/quantizers/native_quantizer.py:139]: Explain why we need to `post_process_weight_fakequant`

File: mmrazor/models/quantizers/native_quantizer.py [hunk 5]
```diff
@@ -96,16 +134,30 @@ def prepare(self, model, graph_module):
...
+                # Trace `SUPPORT_QAT_MODULES` recursively.
                 if isinstance(child, SUPPORT_QAT_MODULES):
+                    # We add w_fakequant once in case some ptq methods have
+                    # specific operations such as Adaround. So we do Quantize
+                    # to perform these operations and do dequantize to
+                    # introduce quantization loss in advance.
                     weight_fakequant = child.weight_fake_quant
                     child.weight.data = weight_fakequant(child.weight.data)
 
+                    # `to_float()` function fuse BN into conv or conv_relu.
```
Comment (pppppM) [inline mmrazor/models/quantizers/native_quantizer.py:156]: `to_float()` is not only fuse BN but also will convert a qat module to a normal module.

Add `to_float()` source code url

File: mmrazor/models/quantizers/native_quantizer.py [hunk 6]
```diff
@@ -96,16 +134,30 @@ def prepare(self, model, graph_module):
...
                     float_child = child.to_float()
 
+                    # This is decided by backend type, some backend need
+                    # explicitly keep the fake quant structure, others don't.
```
Comment (pppppM) [inline mmrazor/models/quantizers/native_quantizer.py:160]:
```suggestion
                    # This is decided by backend type, some backend need
                    # explicitly keep the fake quant structure, others don't.
                    # TODO add deploy doc link
```

File: configs/quantization/qat/base_openvino_resnet18_8*b32_in1k.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
+_base_ = ['mmcls::resnet/resnet18_8xb32_in1k.py']
+
+train_dataloader = dict(batch_size=1024)
+
+# test_cfg = dict(
+#     type='mmrazor.PTQLoop',
+#     calibrate_dataloader=train_dataloader,
+#     calibrate_steps=32,
+# )
+
```
Comment (pppppM) [inline configs/quantization/qat/base_openvino_resnet18_8*b32_in1k.py:10]: del

**Review by pppppM — CHANGES_REQUESTED** [review]
- `configs/quantization/qat/base_openvino_resnet18_8*b32_in1k.py` should be `configs/quantization/qat/minmax_openvino_resnet18_8xb32_in1k.py`.
The first place of the config name should be the algorithm name, and `*` should use `x`.

- Remove `dist_ptq.sh`. We don't plan to support dist ptq
- Complete arg descriptions. Some arg descriptions maybe generated by the plugin,  don't have real context

File: mmrazor/models/algorithms/quantization/mm_architecture.py [hunk 8]
```diff
@@ -138,6 +185,8 @@ def forward(self,
             return self.architecture(inputs, data_samples, mode)
 
     def calibrate_step(self, data):
+        """PTQ method need calibrate by cali data."""
+
         data = self.data_preprocessor(data, False)
         return self._run_forward(data, mode='predict')
```
Comment (humu789) [inline mmrazor/models/algorithms/quantization/mm_architecture.py:192]: MMArchitectureQuantDDP need to add docstring and type hint

File: mmrazor/models/quantizers/native_quantizer.py [hunk 9]
```diff
@@ -56,15 +59,40 @@
...
     # backend: 'native'
     # support_w_modes = ['per_tensor', 'per_channel']
     # support_a_modes = ['per_tensor']
 
     def __init__(self,
-                 global_qconfig,
-                 no_observer_modules=None,
+                 global_qconfig: Union[Dict, Config],
+                 no_observer_modules: Union[List, Tuple, None] = None,
                  tracer=dict(type='CustomTracer'),
```
Comment (humu789) [inline mmrazor/models/quantizers/native_quantizer.py:96]: ` tracer` and `extra_redundant_fakequants` need to add type hint

File: mmrazor/models/quantizers/native_quantizer.py [hunk 10]
```diff
@@ -171,7 +240,6 @@ def prepare_for_mmdeploy(self, model, dummy_input, checkpoint):
         raise NotImplementedError
 
     def del_redundant_fakequant(self, prepared):
```
Comment (humu789) [inline mmrazor/models/quantizers/native_quantizer.py:242]: add docstring and type hint

**Review by humu789 — CHANGES_REQUESTED** [review]
1. need to add some docstring and type hint
2. need to get the coverage of your added unittest

File: mmrazor/models/algorithms/quantization/mm_architecture.py [hunk 11]
```diff
@@ -106,7 +128,31 @@ def traverse(module, prefix):
                 continue
             traverse(self.qmodels[mode], '')
 
-    def _build_qmodels(self, model):
+    def _build_qmodels(self, model: dict):
```
Comment (humu789) [inline mmrazor/models/algorithms/quantization/mm_architecture.py:131]: `model` is not dict

File: mmrazor/models/algorithms/quantization/mm_architecture.py [hunk 12]
```diff
@@ -68,9 +76,23 @@ def __init__(self,
 
         self.qmodels = self._build_qmodels(self.architecture)
 
-        self.sync_qparams('predict')
+        self.sync_qparams(forward_modes[1])
```
Comment (humu789) [inline mmrazor/models/algorithms/quantization/mm_architecture.py:79]: can not use `forward_modes[1]`, `len(forward_modes)` may be is 1 in ptq

File: mmrazor/models/quantizers/native_quantizer.py [hunk 13]
```diff
@@ -56,17 +62,42 @@
...
     # support_w_modes = ['per_tensor', 'per_channel']
     # support_a_modes = ['per_tensor']
 
     def __init__(self,
-                 global_qconfig,
-                 no_observer_modules=None,
-                 tracer=dict(type='CustomTracer'),
-                 extra_redundant_fakequants=dict(
+                 global_qconfig: Union[Dict, Config],
+                 no_observer_modules: Union[List, Tuple, None] = None,
```
Comment (humu789) [inline mmrazor/models/quantizers/native_quantizer.py:98]: `Union[List, Tuple, None]` -> `Optional[List]`

**Review by humu789 — COMMENTED** [review]
`typing` used in type hint and docstring had better be consistent

**Review by humu789 — APPROVED** [review]

**Review by pppppM — APPROVED** [review]

**Pull request merged**